Aws new actions

andrewjschuang opened the pull request:
AWS Components using new [AWS SDK for JavaScript v3](https://docs.aws.amazon.com/AWSJavaScriptSDK/v3/latest/index.html#aws-sdk-for-javascript-v3).
Couldn't find a v3 version for AWS StepFunctions, though it's not a problem since SDK v3 is fully backwards compatible with v2.

New Actions:
- [X] Create Lambda Function
- [X] Send Message to SNS

Refactored Actions (and Common):
- [X] Invoke Lambda Function
- [X] Send Event to EventBridge
- [X] Send Message to SQS
- [X] Stream File to S3
- [X] Upload File to S3
- [X] Put Log Events to CloudWatch Logs

Refactored Sources (and Common):
- [X] New Emails Sent to SES
- [X] New Records Returned by CloudWatch Logs
- [X] New Scheduled Tasks
- [X] New SNS Messages

Please review this along with #2714.

Resolves #1458, Resolves #2150.

dylburger commented:
- [ ] [Create components to address specific use cases whenever possible](https://pipedream.com/docs/components/guidelines/#component-scope)
- [ ] [Component `key`s should follow the format `app_name_slug-slugified-component-name`](https://pipedream.com/docs/components/guidelines/#component-key-pattern)
- [ ] Components should follow the [standard directory structure](https://pipedream.com/docs/components/guidelines/#folder-structure)
- [ ] [Prefer Node.js client libraries to REST APIs](https://pipedream.com/docs/components/guidelines/#using-apis-vs-client-libraries)
- [ ] When making API requests, [handle pagination](https://pipedream.com/docs/components/guidelines/#pagination) to ensure all data / events are processed
- [ ] [Use `secret` props](https://pipedream.com/docs/components/guidelines/#capturing-sensitive-data) to capture sensitive data
- [ ] Props and methods should be defined in [app files](https://pipedream.com/docs/components/guidelines/#app-files) whenever possible
- [ ] Document methods with [JS Docs](https://pipedream.com/docs/components/guidelines/#methods)
- [ ] [Use `optional` props](https://pipedream.com/docs/components/guidelines/#optional-vs-required-props) whenever possible, and [set a `default` value](https://pipedream.com/docs/components/guidelines/#default-values) where you can
- [ ] Use [async options](https://pipedream.com/docs/components/guidelines/#async-options) to accept user input wherever possible
- [ ] [Create subscription webhooks sources (vs polling sources) whenever possible](https://pipedream.com/docs/components/guidelines/#webhook-vs-polling-sources)
- [ ] [Always emit a `summary` with events](https://pipedream.com/docs/components/guidelines/#emit-a-summary)
- [ ] [Use a built-in deduping strategy wherever possible](https://pipedream.com/docs/components/guidelines/#deduping)
- [ ] When building webhook sources, [follow these best practices](https://pipedream.com/docs/components/guidelines/#webhook-sources)
- [ ] When building polling sources, [follow these best practices](https://pipedream.com/docs/components/guidelines/#polling-sources)

andrewjschuang commented on components/aws/actions/cloudwatch-logs-put-log-event/cloudwatch-logs-put-log-event.mjs:
```diff
@@ -0,0 +1,86 @@
+    Uploads a log event to the specified log stream.
+    [See docs](https://docs.aws.amazon.com/AWSJavaScriptSDK/v3/latest/clients/client-cloudwatch-logs/classes/putlogeventscommand.html)
+  `),
+  version: "0.0.2",
```
I changed some action's keys to follow a structure, `aws-<aws_service>-<action>`.
I'm not sure if the version should go back to `0.0.1` or be bumped from the previous number.

andrewjschuang commented on components/aws/actions/dynamodb-put-item/dynamodb-put-item.mjs:
```diff
@@ -1,118 +1,99 @@
+          description: "Value for the sort key",
+        };
+      }
+      props.item = {
```
`Item` doesn't need to be an additional prop, but when put here it maintains a better prop order in the UI. A feature for ordering props would be great for these cases

andrewjschuang commented on components/aws/common/common.mjs:
```diff
@@ -0,0 +1,124 @@
+    eventBridge: EventBridgeClient,
+    sqs: SQSClient,
+    sns: SNSClient,
+    ses: SESClient,
```
For some reason if `SESClient` is kept, when publishing any action pipedream api will return http 500

js07 replied:
I see the same error when publishing components. Removing several of the other awsClients also fixes the 500 error for me. I think the issue might be that each component's bundle size is too large to publish with all of the packages included.

andrewjschuang replied:
Yeah, I tried that and got the same results. What should we do in this case?

js07 replied:
To reduce the bundle size, we could remove the (component-specific) AWS SDK client imports (and methods) from the app file–which is imported in every component–and import them in only components that use that client. And we could create a common file for each AWS service/client like the [common files](https://github.com/andrewjschuang/pipedream/tree/0081cdbe9a6cc3ede53218994cf0a368665b132e/components/aws/sources/common) for the AWS sources. The common files would import the SDK packages and instantiate the clients using a method in the app file. Let me know if you think that'll work.

andrewjschuang commented on components/aws/common/constants.mjs:
```diff
@@ -0,0 +1,58 @@
+};
+
+const dynamodb = {
+  keyTypes: createObjectFromArray([
```
I was thinking if this way of creating a constant is better than the above (`clients`). It is cleaner imo, but there is code completion with the `clients` constant.

andrewjschuang replied:
Yeah, I agree with you. Addressed

js07 commented on components/aws/common/common.mjs:
```diff
@@ -0,0 +1,124 @@
+
+export default {
+  defaultRegion: "us-east-1",
+  awsClients: {
```
By convention, files with the name `common.mjs` in this repo are used like base classes, with the same structure as components. (See the [guideline docs](https://pipedream.com/docs/components/guidelines/#common-files-optional))

I'd suggest renaming this file, or splitting this file into a `regions.mjs` file and `clients.mjs` file?

andrewjschuang replied:
Addressed

js07 reviewed (changes requested):
@andrewjschuang The refactored components and new actions look great overall! I made some minor comments.

andrewjschuang commented on components/aws/actions/cloudwatch-logs-put-log-event/cloudwatch-logs-put-log-event.mjs:
```diff
@@ -0,0 +1,86 @@
+
+export default {
+  key: "aws-cloudwatch-logs-put-log-event",
+  name: "AWS - CloudWatch Logs - Put Log Event",
```
Addressed

andrewjschuang commented on components/aws/actions/dynamodb-execute-statement/dynamodb-execute-statement.mjs:
```diff
@@ -1,57 +1,50 @@
 
-    //Sends the request using the DynamoDB AWS object
-    return await dynamodb.executeTransaction(dynamoDbParams).promise();
+    $.export("$summary", `Statement returned ${response.Items.length} items`);
```
Addressed

andrewjschuang commented on components/aws/actions/dynamodb-put-item/dynamodb-put-item.mjs:
```diff
@@ -1,118 +1,99 @@
-      ExpressionAttributeValues: this.expression_attribute_values,
+    params.Item = {
+      ...params.Item,
+      ...JSON.parse(this.item || "{}"),
```
Addressed

andrewjschuang commented on components/aws/actions/lambda-create-function/lambda-create-function.mjs:
```diff
@@ -0,0 +1,53 @@
+        aws,
+        "region",
+      ],
+      description: "The AWS region tied to your Lambda, e.g us-east-1 or us-west-2",
```
Addressed

andrewjschuang commented on components/aws/actions/lambda-invoke-function/lambda-invoke-function.mjs:
```diff
@@ -0,0 +1,54 @@
+        aws,
+        "region",
+      ],
+      description: "The AWS region tied to your Lambda, e.g us-east-1 or us-west-2",
```
Addressed

andrewjschuang commented on components/aws/actions/sns-send-message/sns-send-message.mjs:
```diff
@@ -0,0 +1,42 @@
+        aws,
+        "region",
+      ],
+      description: "Region tied to your SNS Topic, e.g. us-east-1 or us-west-2",
```
Addressed

andrewjschuang commented on components/aws/actions/sqs-send-message/sqs-send-message.mjs:
```diff
@@ -0,0 +1,43 @@
+        aws,
+        "region",
+      ],
+      description: "The AWS region tied to your SQS queue, e.g us-east-1 or us-west-2",
```
Addressed

andrewjschuang commented on components/aws/aws.app.mjs:
```diff
@@ -1,42 +1,209 @@
+    role: {
+      type: "string",
+      label: "Role",
+      description: "An IAM Role ARN, e.g., arn:aws:iam::650138640062:role/v3-lambda-tutorial-lambda-role. [See docs](https://docs.aws.amazon.com/lambda/latest/dg/lambda-intro-execution-role.html)",
```
Addressed

andrewjschuang commented on components/aws/aws.app.mjs:
```diff
@@ -71,66 +238,133 @@ export default {
+    streamSpecificationEnabled: {
+      type: "boolean",
+      label: "Stream Specification Enabled",
+      description: "Indicates whether DynamoDB Streams is to be enabled (true) or disabled (false)",
```
Addressed

andrewjschuang commented on components/aws/actions/cloudwatch-logs-put-log-event/cloudwatch-logs-put-log-event.mjs:
```diff
@@ -0,0 +1,86 @@
+    Uploads a log event to the specified log stream.
+    [See docs](https://docs.aws.amazon.com/AWSJavaScriptSDK/v3/latest/clients/client-cloudwatch-logs/classes/putlogeventscommand.html)
+  `),
+  version: "0.0.2",
```
Addressed

andrewjschuang commented on components/aws/actions/dynamodb-update-table/dynamodb-update-table.mjs:
```diff
@@ -1,97 +1,98 @@
+    Modifies the settings for a given table. Only one type of modification is permitted per request.
+    [See docs](https://docs.aws.amazon.com/AWSJavaScriptSDK/v3/latest/clients/client-dynamodb/classes/updatetablecommand.html)
+  `),
+  version: "0.1.10",
```
Addressed

andrewjschuang commented on components/aws/aws.app.mjs:
```diff
@@ -1,42 +1,209 @@
+      description: "An IAM Role ARN, e.g., arn:aws:iam::650138640062:role/v3-lambda-tutorial-lambda-role. [See docs](https://docs.aws.amazon.com/lambda/latest/dg/lambda-intro-execution-role.html)",
+      async options({ region }) {
+        const response = await this.listRoles(region);
+        return response.Roles.map((role) => role.Arn);
```
Addressed

andrewjschuang commented on components/aws/actions/dynamodb-execute-statement/dynamodb-execute-statement.mjs:
```diff
@@ -1,57 +1,50 @@
-      secretAccessKey,
-      region: this.region,
-    });
+    if (this.parameters) params.Parameters = this.parameters;
```
If `params.Parameters` is `undefined`, the request will **not** fail. I changed it to the following:
```
if (this.parameters?.length > 0) {
  const p = this.parameters.filter((param) => param);
  if (p.length > 0) params.Parameters = p;
}
```
This is because props that are an `array` of `strings` may contain empty strings; Pipdream doesn't strip them. If there are empty strings the API request will fail

andrewjschuang commented on components/aws/actions/dynamodb-update-table/dynamodb-update-table.mjs:
```diff
@@ -1,97 +1,98 @@
+      params.StreamSpecification = {
+        StreamEnabled: this.streamSpecificationEnabled,
+      };
+      if (this.streamSpecificationViewType) {
+        params.StreamSpecification.StreamViewType = this.streamSpecificationViewType;
+      }
```
Addressed

andrewjschuang commented:
@js07 Thanks for the review!

I addressed most issues but in a few of them I've asked some questions in the comments

js07 reviewed (changes requested):
@andrewjschuang The changes look good. I made a few minor comments and commented on the [discussion](https://github.com/PipedreamHQ/pipedream/pull/2302#discussion_r825231835) above.

andrewjschuang commented on components/aws/aws.app.mjs:
```diff
@@ -401,8 +408,14 @@ export default {
+      return response.Table.AttributeDefinitions;
+    },
+    decodeResponsePayload(payload) {
+      return JSON.parse(new TextDecoder("utf-8").decode(payload) || {});
```
Addressed

andrewjschuang commented on components/aws/common/common.mjs:
```diff
@@ -0,0 +1,124 @@
+import { SESClient } from "@aws-sdk/client-ses";
+import { DynamoDBClient } from "@aws-sdk/client-dynamodb";
+
+export default {
```
Addressed

andrewjschuang commented on components/aws/aws.app.mjs:
```diff
@@ -71,66 +245,133 @@ export default {
-     * by calling the EC2 `DescribeRegions` API.
-     * @returns A new configured instance of the AWS SDK
-     */
-    sdk(region) {
```
Addressed

jcortes reviewed (changes requested):
Hi @andrewjschuang everything looks good to me, just a minor request change. Make sure you don't have any conflict errors and it will be good for QA Ready

andrewjschuang commented:
> Hi @andrewjschuang everything looks good to me, just a minor request change. Make sure you don't have any conflict errors and it will be good for QA Ready

Hey @jcortes, I fixed the conflicts and also refactored `aws-new-emails-sent-to-ses-catch-all-domain` source to use AWS v3 clients

jcortes reviewed (approved):
Hi @andrewjschuang this one LGTM, nice work!

dannyroosevelt commented:
@andrewjschuang it looks like you refactored all of the existing `aws` actions, including those for DynamoDB, is that right?

andrewjschuang commented:
> @andrewjschuang it looks like you refactored all of the existing aws actions, including those for DynamoDB, is that right?

@dannyroosevelt yes, in this PR these are the updated actions:

 - [X] cloudwatch-logs-put-log-event
 - [X] dynamodb-create-table
 - [X] dynamodb-execute-statement
 - [X] dynamodb-get-item
 - [X] dynamodb-put-item
 - [X] dynamodb-query
 - [X] dynamodb-scan
 - [X] dynamodb-update-item
 - [X] dynamodb-update-table
 - [X] eventbridge-send-event
 - [X] lambda-create-function
 - [X] lambda-invoke-function
 - [X] s3-stream-file
 - [X] s3-upload-file
 - [X] sns-send-message
 - [X] sqs-send-message

And sources:

 - [X] new-emails-sent-to-ses-catch-all-domain
 - [X] new-records-returned-by-cloudwatch-logs-insights-query
 - [X] new-scheduled-tasks
 - [X] new-sns-messages

andrewjschuang commented:
Hey @jcortes can you approve this just so that I can merge this PR? There were some changes in the code so GitHub removes the approved label
